### DISCUSSION TITLE
Static compilation

### DISCUSSION THREAD

**Discussion by manio** [zmwangx/rust-ffmpeg#87] [Q&A]
Hi!
I am wondering if there is some easy way to create a static build (with all ffmpeg functions inside) of a rust executable.
When I am compiling a sample test program which is using `ffmpeg-next` then the `ldd` on the result file (on target host) is:

```
$ ldd ./ffmpeg-test
	linux-vdso.so.1 =>  (0x00007ffcfd719000)
	libavutil.so.55 => not found
	libavformat.so.57 => not found
	libavfilter.so.6 => not found
	libavdevice.so.57 => not found
	libswscale.so.4 => not found
	libswresample.so.2 => not found
	libavcodec.so.57 => not found
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fe56e6d2000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fe56e4ca000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fe56e2ae000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fe56e0aa000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fe56dcdc000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fe56eb40000)
```
As the result I have to install ffmpeg. Is it possible to create a static binary with ffmpeg libs included in a result binary?

**Comment by Ticsmtc** [comment]
Hi there , it’s been a while since i doing this project , pls wait I will check my build , make sure I get static executable. : )

**Comment by Ticsmtc** [comment]
Yes , I checked my project build , it’s static by default . I am using this tool build my project and it’s static by default.

https://github.com/rust-embedded/cross <https://github.com/rust-embedded/cross>

I wish I can help you , but that’s all I get .

Also I use -vv option check out rustc args and I found this line during my build : cargo:rustc-link-lib=static=global_static

Check out the cross project I mention above , at least make you cross-building easier .

Wish you best.

**Reply by manio** [reply]
Thanks for your tips. I don't need to crosscompile my project. I want to just compile for the same architecture but include missing libs which can be missing on target machine.

I was able to tell compiler to include eg. libavcodec with this in `build.rs`:
```println!("cargo:rustc-link-lib=avcodec");```
But the resulting binary still have libavcodec when I do `ldd` on it...

Just curious: can you output the ldd of your binary?

**Reply by Ticsmtc** [reply]
i use `readelf` to make sure i have static build , I dont have native arm32 ldd right now .

 try this 
`println!("cargo:rustc-link-lib=static=avcodec");` in build.sh force cargo use static library maybe help.

**Reply by manio** [reply]
I was also trying this but then it is telling me that:
```= note: /usr/bin/ld: cannot find -lavcodec```
So I added the library patch with:
```println!(r"cargo:rustc-link-search=\lib64\");```
But still this problem...

**Reply by Ticsmtc** [reply]
you sure you have the static library ? eg. libavcodec.a in path ?

**Reply by manio** [reply]
There is a shared libs. Should there be .a for static linking ?
```bash
$ ll /lib64/libavco*
lrwxrwxrwx. 1 root root       24 Jan  7 06:21 /lib64/libavcodec.so -> libavcodec.so.57.107.100
lrwxrwxrwx. 1 root root       24 Jan  7 06:20 /lib64/libavcodec.so.57 -> libavcodec.so.57.107.100
-rwxr-xr-x. 1 root root 11747248 Oct 13 04:01 /lib64/libavcodec.so.57.107.100
```

**Comment by Ticsmtc** [comment]
yes , you must have .a for static link , i get
```bash
-rw-r--r-- 1 root root 89252710 Dec 20 14:24 /usr/local/lib/libavcodec.a
lrwxrwxrwx 1 root root       23 Dec 20 14:24 /usr/local/lib/libavcodec.so -> libavcodec.so.58.91.100
lrwxrwxrwx 1 root root       23 Dec 20 14:24 /usr/local/lib/libavcodec.so.58 -> libavcodec.so.58.91.100
-rwxr-xr-x 1 root root 47078232 Dec 20 14:24 /usr/local/lib/libavcodec.so.58.91.100
```

if you use system package manager ( apt , yum ) install your ffmpeg , there may be just shared library , or you can try to install dev-package , eg. ffmpeg-dev or something , depend on your linux system.

if you compile ffmpeg yourself , you may add `--enable-static` to your `configrue` step 
`./configure --enable-static`

i think your problem can be solved  : )
